src/envoy/http/authn/http_filter.cc

```diff
@@ -30,59 +57,207 @@ void AuthenticationFilter::onDestroy() {
+          // On the other hand, it's ok to have no (origin) methods if
+          // rule USE_SOURCE
+          onAuthenticateOriginDone(headers, &rule, 0, nullptr, true);
+          break;
+        default:
+          // Should never come here.
+          ENVOY_LOG(error, "Invalid binding value for rule {}",
+                    rule.DebugString());
+          break;
       }
+      return;
+    }
+    authenticateOrigin(
+        *headers, rule.origins(0),
+        std::bind(&AuthenticationFilter::onAuthenticateOriginDone, this,
+                  headers, &rule, 0, std::placeholders::_1,
+                  std::placeholders::_2));
+  }
+}
+
+void AuthenticationFilter::authenticateOrigin(
+    HeaderMap& headers, const iaapi::OriginAuthenticationMethod& method,
+    const AuthenticateDoneCallback& done_callback) {
+  validateJwt(headers, method.jwt(), done_callback);
+}
+
+void AuthenticationFilter::onAuthenticateOriginDone(
+    HeaderMap* headers, const iaapi::CredentialRule* rule, int method_index,
+    std::unique_ptr<IstioAuthN::AuthenticatePayload> payload, bool success) {
+  if (!success) {
+    // Authentication fail, try the next method, if available.
+    method_index++;
+    if (method_index < rule->origins_size()) {
+      authenticateOrigin(
+          *headers, rule->origins(method_index),
+          std::bind(&AuthenticationFilter::onAuthenticateOriginDone, this,
+                    headers, rule, method_index, std::placeholders::_1,
+                    std::placeholders::_2));
     } else {
-      ENVOY_LOG(
-          debug,
-          "AuthenticationFilter: {} this connection does not require mTLS",
-          __func__);
+      rejectRequest("Origin authentication failed.");
+    }
+  } else {
+    // Authentication pass, look at the return payload and store to the context
+    // output. Set filter to continueDecoding when done.
+    // At the moment, only JWT can be used for origin authentication, so
+    // it's ok just to check jwt payload.
+    if (payload != nullptr && payload->has_jwt()) {
+      *context_.mutable_origin() = payload->jwt();
+    }
+    switch (rule->binding()) {
+      case iaapi::CredentialRule::USE_PEER:
+        context_.set_principal(context_.peer_user());
+        break;
+      case iaapi::CredentialRule::USE_ORIGIN:
+        context_.set_principal(context_.origin().user());
+        break;
+      default:
+        // Should never come here.
+        ENVOY_LOG(error, "Invalid binding value for rule {}",
+                  rule->DebugString());
+        break;
     }
+    // It's done. continueDecoding to accept request.
+    continueDecoding();
   }
+}
 
-  ENVOY_LOG(
-      debug,
-      "Called AuthenticationFilter : {}, return FilterHeadersStatus::Continue;",
-      __func__);
-  return FilterHeadersStatus::Continue;
+void AuthenticationFilter::continueDecoding() {
+  state_ = IstioAuthN::State::COMPLETE;
+  if (stopped_) {
+    decoder_callbacks_->continueDecoding();
```

Comment:
not sure if continueDecoding can be re-entry,  or can be called multiple times for a filter.

Reply:
It is called once. The flow will call this function only when it doesn't need to perform authN any further.